lib/ddtrace/contrib/sinatra/tracer_middleware.rb

```diff
@@ -1,64 +1,61 @@
-require 'ddtrace/contrib/sinatra/request_span'
+require 'ddtrace/contrib/sinatra/env'
+require 'ddtrace/contrib/sinatra/headers'
 
 module Datadog
   module Contrib
     module Sinatra
       # Middleware used for automatically tagging configured headers and handle request span
       class TracerMiddleware
+        REQUEST_TRACE_NAME = 'sinatra.request'.freeze
+
         def initialize(app)
           @app = app
         end
 
         def call(env)
-          span = RequestSpan.span!(env)
+          # Extend the Env with Sinatra tracing functions
+          env.extend(Sinatra::Env)
```

Comment:
Interesting approach. I'm worried however that this might introduce somewhat unstable API to rack envs.

Imagine another rack middleware used in the same application. That does use `env.datadog_span` accessor methods. It all works well as long as the order of MWs is `TracerMiddleware -> OtherMiddleware`. 

When the order gets reversed then calling `env.datadog_span` will throw `NoMethodError`. 

As such I think it would be much safer to use module functions to share the code operating on `env`. 
If avoiding object allocations wasn't one of our goals, I would suggest using Composition over Inheritance. Because `env.extend` is Ruby specific runtime inheritance.

In general also extending objects this way should be avoided due to reasons mentioned above, where order of execution change would mean the code could suddenly throw `NoMethodError`.

Reply:
Fair concern about `datadog_span`, regarding the change in middleware order. However, I don't think whether you `extend` or call the module will make a difference anyways.

The reason for saying so is imagine this second middleware that was dependent on `env.datadog_span`. Instead, it will call to a function like `Sinatra::Env.datadog_span(env)` which won't raise a `NoMethodError` now, but will still fail because the span will be missing from the `env` because `TracerMiddleware` didn't run yet. In this case, `OtherMiddleware` still depends on `TracerMiddleware` no matter how you articulate it. And if there is such a dependency, it's reasonable that the dependent be responsible for placing itself in the right position (i.e. the developer should be responsible for configuring their middleware stack correctly.)

However,  I will yield that `extend` is a bit unorthodox, and by using it, you might have inconsistency between objects of the same type; something that might not sit well with many developers. As such, I think it is reasonable to remove the `extend` implementation in favor of something simpler, albeit not as elegant.